COMPATIBILITY.md

```diff
@@ -0,0 +1,102 @@
+release had incompatible changes that broke `rules_rust`), it is acceptable to
+merge PRs leaving the build red as long as the reason for failure remains the
+same. We hope red CI with the current rolling release will be rare.
+
+`rules_rust` don't promise all new `rules_rust` features available to the
+current Bazel rolling release to be be available to the current Bazel LTS
+release (because Bazel compatibility policy doesn't allow us to make that
+promise, and some new features of `rules_rust` require new Bazel features that
+are only available in Bazel rolling releases). `rules_rust` do promise that new
+features available to the current Bazel rolling release will be available to the
+next Bazel LTS release at latest.
+
+Whenever there is a new Bazel LTS release, all releases of `rules_rust` in the
+first 3 months will support both Bazel LTS versions.
+
+## What host platforms are supported?
+
+The only platform subject to backwards compatibility policy is `Linux` on
+`x86_64`, as the only fully supported platform. Process for moving a best effort
+platform to a supported platform is consensus-based.
+
+## What are the stable APIs of `rules_rust`?
+
+`//rust:defs.bzl` is subject to the backwards compatibility policy. That means
+that anything directly accessible from this file is considered stable.
+
+`//rust/private/…` is not subject to the backwards compatibility policy. Content
+of this package is an implementation detail.
+
+`//cargo`, `//util`, `//tools`, `//test`, `//examples`, `//bindgen`, `//proto`
+and any packages not mentioned by this document are by default not subject to
+the backwards compatibility policy.
+
+Experimental build settings are not subject to the backward compatibility
+policy. They should be added to `//rust:experimental.bzl`.
+
+Incompatible build settings are subject to the backward compatibility policy,
+meaning the behavior of the flag cannot change in a backwards incompatible way.
+They should be added to `//rust:incompatible.bzl`.
+
+Bug fixes are not a breaking change by default. We'll use Common Sense (and we
+will pull in more maintainers and the community to discuss) if we see a certain
+bug fix is controversial.
+
+## How to make a backwards incompatible change?
+
+1. Create a GitHub issue (example:
+[Split rust\_library into smaller rules#591](https://github.com/bazelbuild/rules_rust/issues/591)).
+2. Describe the change, motivation for the change, provide migration
+instructions/tooling.
+3. Add a build setting into `//rust:incompatible.bzl` that removes the old
+behavior (whenever possible) or changes the current behavior (when just
+removing old behavior is not possible). Users should not need to manually
+flip incompatible flags. If for a specific flag they do need to manually flip
+it, then in the first backwards incompatible release we make the flag be a
+noop, and only in the second backwards incompatible release we remove the
```

Comment:
Is the idea that the value of the flag that was the _old_ default will now error, but if it's set to the _new_ default it will no-op?

i.e. if we introduced`--incompatible_disable_borrowck=False`, the incompatible release would default `--incompatible_disable_borrowck=True` and setting `--incompatible_disable_borrowck=False` would error?

(The phrase "the flag would be a noop" suggests to me that there would be no error, and that whether you set it to `True` or `False` you'd get the same value)

Reply:
So `--incompatible_disable_borrowck` is a well behaving flag, you can migrate your repo (= stop relying on borrowck) without having to flip the flag. Because you don't have to flip it, you won't, and therefore we can remove the flag silently when it is not needed (without changing the default value of the flag).

If you had a flag like `--incompatible_use_new_orphan_rule` that forces you to either have old behavior with the flag set to false, or new behavior with flag set to true, but not new behavior with flag set to false, only then we would have to first have a release when the behavior in rules is hardcoded for the new orphan rule, but the flag still exists but does nothing (so you can simply remove it from your bazelrc). Am I making any sense? Can you think of a different wording that is clearer?

Reply:
I think the bit I'm not clear on is that I'm not sure what the value is in keeping the flag around when it doesn't do anything? Why not delete it as soon as the old behaviour is no longer supported?

Reply:
Imagine user migrated for `incompatible_use_new_orphan_rule`, and they added the flag to their bazelrc. If we remove the flag, their build will break.

But thinking about it now, it is a bit too complicated, and the value provided is rather small. I'll remove this bit from the text.